File: implementations/node/core/years/2022/day-5/part-1.mjs
import { executeInstructions, parseInput } from "./common.mjs";

export function execute(input) {
    const [columns, instructions] = parseInput(input)
    return executeInstructions(columns, instructions, (columns, [n, from, to]) => {
        for(let i = 0; i < n; i++) {
            columns[to].push(columns[from].pop())
        }
    })
}

File: implementations/rust/src/years/y2022/day_5/part_1.rs
use crate::years::y2022::day_5::common::{parse_input, take_tops};

fn execute_commands(columns: &mut Vec<Vec<char>>, instructions: Vec<[usize; 3]>) {
    instructions.into_iter().for_each(|[n, from, to]| {
        for _ in 0..n {
            let top = columns[from].pop().expect("Top from column");
            columns[to].push(top);
        }
    })
}

pub fn execute(input: String) -> String {
    let (mut columns, instructions) = parse_input(&input);

    execute_commands(&mut columns, instructions);

    return take_tops(&columns).iter().collect();
}